File: csharp/patterns/src/PrototypePatterns/Creational/Singleton/SingletonPattern.cs
namespace PrototypePatterns.Creational.Singleton;

public sealed class Configuration
{
    private static readonly Configuration InstanceValue = new();
    private readonly Dictionary<string, string> _settings = new();

    private Configuration()
    {
    }

    public static Configuration Instance => InstanceValue;

    public void Set(string key, string value)
    {
        _settings[key] = value;
    }

    public string? Get(string key)
    {
        return _settings.TryGetValue(key, out var value) ? value : null;
    }
}

public static class SingletonPattern
{
    public static bool Demo()
    {
        var first = Configuration.Instance;
        var second = Configuration.Instance;
        first.Set("mode", "prototype");
        return ReferenceEquals(first, second) && second.Get("mode") == "prototype";
    }
}

File: java/patterns/src/main/java/org/prototypepatterns/creational/singleton/SingletonPattern.java
package org.prototypepatterns.creational.singleton;

import java.util.HashMap;
import java.util.Map;

public final class SingletonPattern {
    private SingletonPattern() {
    }

    public static final class Configuration {
        private static final Configuration INSTANCE = new Configuration();
        private final Map<String, String> settings = new HashMap<>();

        private Configuration() {
        }

        public static Configuration instance() {
            return INSTANCE;
        }

        public void set(String key, String value) {
            settings.put(key, value);
        }

        public String get(String key) {
            return settings.get(key);
        }
    }

    public static boolean demo() {
        Configuration first = Configuration.instance();
        Configuration second = Configuration.instance();
        first.set("mode", "prototype");
        return first == second && "prototype".equals(second.get("mode"));
    }
}
